pkg/network/node/metrics.go

```diff
@@ -0,0 +1,192 @@
+	SDNSubsystem = "sdn"
+
+	OVSFlowsKey                 = "ovs_flows"
+	ARPCacheAvailableEntriesKey = "arp_cache_entries"
+	PodIPsKey                   = "pod_ips"
+	PodSetupErrorsKey           = "pod_setup_errors"
+	PodSetupLatencyKey          = "pod_setup_latency"
+	PodTeardownErrorsKey        = "pod_teardown_errors"
+	PodTeardownLatencyKey       = "pod_teardown_latency"
+)
+
+var (
+	OVSFlows = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: SDNNamespace,
+			Subsystem: SDNSubsystem,
+			Name:      OVSFlowsKey,
+			Help:      "Number of Open vSwitch flows",
+		},
+	)
+
+	ARPCacheAvailableEntries = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: SDNNamespace,
+			Subsystem: SDNSubsystem,
+			Name:      ARPCacheAvailableEntriesKey,
+			Help:      "Number of available entries in the ARP cache",
+		},
+	)
+
+	PodIPs = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: SDNNamespace,
+			Subsystem: SDNSubsystem,
+			Name:      PodIPsKey,
+			Help:      "Number of allocated pod IPs",
+		},
+	)
+
+	PodSetupErrors = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: SDNNamespace,
+			Subsystem: SDNSubsystem,
+			Name:      PodSetupErrorsKey,
+			Help:      "Number pod setup errors",
+		},
+	)
+
+	PodSetupLatency = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace: SDNNamespace,
+			Subsystem: SDNSubsystem,
+			Name:      PodSetupLatencyKey,
+			Help:      "Latency of pod network setup in microseconds",
+		},
+		[]string{"pod_namespace", "pod_name", "sandbox_id"},
+	)
+
+	PodTeardownErrors = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: SDNNamespace,
+			Subsystem: SDNSubsystem,
+			Name:      PodTeardownErrorsKey,
+			Help:      "Number pod teardown errors",
+		},
+	)
+
+	PodTeardownLatency = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace: SDNNamespace,
+			Subsystem: SDNSubsystem,
+			Name:      PodTeardownLatencyKey,
+			Help:      "Latency of pod network teardown in microseconds",
+		},
+		[]string{"pod_namespace", "pod_name", "sandbox_id"},
+	)
+
+	// num stale OVS flows (flows that reference non-existent ports)
+	// num vnids (in the master)
+	// num netnamespaces (in the master)
+	// iptables call time (in upstream kube)
+	// iptables call failures (in upstream kube)
+	// iptables num rules (in upstream kube)
+)
+
+var registerMetrics sync.Once
+
+// Register all node metrics.
+func RegisterMetrics() {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(OVSFlows)
+		prometheus.MustRegister(ARPCacheAvailableEntries)
+		prometheus.MustRegister(PodIPs)
+		prometheus.MustRegister(PodSetupErrors)
+		prometheus.MustRegister(PodSetupLatency)
+		prometheus.MustRegister(PodTeardownErrors)
+		prometheus.MustRegister(PodTeardownLatency)
+	})
+}
+
+// Gets the time since the specified start in microseconds.
+func sinceInMicroseconds(start time.Time) float64 {
+	return float64(time.Since(start).Nanoseconds() / time.Microsecond.Nanoseconds())
```

Comment:
Overly complicated:

```
float64(time.Since(start) / time.Microsecond)
```

is the same thing.  Noticed while reviewing the metrics on a running cluster.

Reply:
@smarterclayton necro-ing this PR, it's actually not the same thing.  gotta go back to the original with Nanoseconds()...